File: core/src/main/java/org/toml/lang/psi/TomlStringLiteralManipulator.java
/*
 * Use of this source code is governed by the MIT license that can be
 * found in the LICENSE file.
 */

package org.toml.lang.psi;

import consulo.annotation.component.ExtensionImpl;
import consulo.document.util.TextRange;
import consulo.language.psi.AbstractElementManipulator;
import jakarta.annotation.Nonnull;
import jakarta.annotation.Nullable;
import org.toml.lang.psi.ext.TomlLiteralExt;
import org.toml.lang.psi.ext.TomlLiteralKind;

@ExtensionImpl
public class TomlStringLiteralManipulator extends AbstractElementManipulator<TomlLiteral> {

    @Nullable
    @Override
    public TomlLiteral handleContentChange(@Nonnull TomlLiteral element, @Nonnull TextRange range, String newContent) {
        String oldText = element.getText();
        String newText = oldText.substring(0, range.getStartOffset()) + newContent + oldText.substring(range.getEndOffset());

        TomlLiteral newLiteral = new TomlPsiFactory(element.getProject()).createLiteral(newText);
        return (TomlLiteral) element.replace(newLiteral);
    }

    @Nonnull
    @Override
    public TextRange getRangeInElement(@Nonnull TomlLiteral element) {
        TomlLiteralKind kind = TomlLiteralExt.getKind(element);
        if (kind instanceof TomlLiteralKind.StringKind) {
            TextRange value = ((TomlLiteralKind.StringKind) kind).getOffsets().getValue();
            if (value != null) {
                return value;
            }
        }
        return super.getRangeInElement(element);
    }

    @Nonnull
    @Override
    public Class<TomlLiteral> getElementClass() {
        return TomlLiteral.class;
    }
}

File: core/src/main/kotlin/org/toml/lang/psi/TomlStringLiteralManipulator.kt
/*
 * Use of this source code is governed by the MIT license that can be
 * found in the LICENSE file.
 */

package org.toml.lang.psi

import consulo.annotation.component.ExtensionImpl
import consulo.document.util.TextRange
import consulo.language.psi.AbstractElementManipulator
import org.toml.lang.psi.ext.TomlLiteralKind
import org.toml.lang.psi.ext.kind

@ExtensionImpl
class TomlStringLiteralManipulator : AbstractElementManipulator<TomlLiteral>() {
    override fun handleContentChange(element: TomlLiteral, range: TextRange, newContent: String?): TomlLiteral {
        val oldText = element.text
        val newText = "${oldText.substring(0, range.startOffset)}$newContent${oldText.substring(range.endOffset)}"

        val newLiteral = TomlPsiFactory(element.project).createLiteral(newText)
        return element.replace(newLiteral) as TomlLiteral
    }

    override fun getRangeInElement(element: TomlLiteral): TextRange {
        return (element.kind as? TomlLiteralKind.String)?.offsets?.value ?: super.getRangeInElement(element)
    }

    override fun getElementClass(): Class<TomlLiteral> {
        return TomlLiteral::class.java
    }
}
